File: app/src/main/java/it/gruppoinfor/home2work/utils/SessionManager.java
package it.gruppoinfor.home2work.utils;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.firebase.iid.FirebaseInstanceId;

import it.gruppoinfor.home2work.App;
import it.gruppoinfor.home2workapi.interfaces.LoginCallback;


/**
 * Created by Federico on 10/01/2017.
 * Permette di salvare la sessione dell'utente nelle preferenze per un login rapido
 */

public class SessionManager {

    public static final int EXPIRED_TOKEN = 0;
    public static final int ERROR = 1;
    public static final int NO_INTERNET = 2;
    public static final String AUTH_CODE = "auth_code";
    private static final String PREFS_SESSION = "it.fleetup.app.session";
    private final SharedPreferences prefs;
    private final String KEY_TOKEN = "token";
    private final String KEY_EMAIL = "email";
    private SessionManagerCallback mCallback;
    private LoginCallback mLoginCallback = new LoginCallback() {
        @Override
        public void onLoginSuccess() {
            storeSession();
            mCallback.onValidSession();
        }

        @Override
        public void onInvalidCredential() {
            mCallback.onExpiredToken();
            signOutUser();
        }

        @Override
        public void onLoginError() {
            mCallback.onExpiredToken();
        }

        @Override
        public void onError(Throwable throwable) {
            mCallback.onError(throwable);
        }
    };

    public SessionManager(Context context) {
        prefs = context.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE);
    }

    public SessionManager(Context context, SessionManagerCallback sessionManagerCallback) {
        this(context);
        mCallback = sessionManagerCallback;
    }

    public void storeSession() {
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_EMAIL, App.home2WorkClient.getUser().getEmail());
        editor.putString(KEY_TOKEN, App.home2WorkClient.getUser().getToken());
        editor.apply();

        String token = FirebaseInstanceId.getInstance().getToken();

        App.home2WorkClient.setFcmToken(token);
    }

    public void loadSession() {

        if (mCallback == null) throw new IllegalStateException("Callback non implementati");

        if (App.home2WorkClient.getUser() != null) {
            mCallback.onValidSession();
            return;
        }

        String email = prefs.getString(KEY_EMAIL, null);
        String token = prefs.getString(KEY_TOKEN, null);

        if (email == null && token == null) {
            mCallback.onNoSession();
            return;
        }

        App.home2WorkClient.login(email, token, true, mLoginCallback);

    }

    public void signOutUser() {
        SharedPreferences.Editor editor = prefs.edit();
        editor.clear();
        editor.apply();
    }

    public interface SessionManagerCallback {

        void onNoSession();

        void onValidSession();

        void onExpiredToken();

        void onError(Throwable throwable);
    }

}

File: App/src/main/java/it/gruppoinfor/home2work/user/SessionManager.kt
package it.gruppoinfor.home2work.user

import android.content.Context
import android.content.Intent
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.LoginEvent
import it.gruppoinfor.home2work.services.LocationService
import it.gruppoinfor.home2work.services.SyncService
import it.gruppoinfor.home2work.utils.Const
import it.gruppoinfor.home2workapi.HomeToWorkClient
import it.gruppoinfor.home2workapi.auth.AuthCallback
import it.gruppoinfor.home2workapi.auth.AuthUser
import it.gruppoinfor.home2workapi.user.User
import java.net.UnknownHostException


/**
 * Created by Federico on 10/01/2017.
 * Permette di salvare la sessione dell'utente nelle preferenze e di ripristinarla all'occorrenza (se non scaduta)
 */

object SessionManager {

    private const val PREFS_SESSION = "it.home2work.app.session"
    private const val PREFS_TOKEN = "ACCESS_TOKEN"

    fun storeSession(ctx: Context, user: AuthUser?) {
        if (user == null) return

        val prefs = ctx.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE)
        val editor = prefs.edit()

        editor.putString(PREFS_TOKEN, user.accessToken)
        editor.apply()

    }

    fun loadSession(ctx: Context, callback: SessionCallback) {
        val prefs = ctx.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE)

        val user: User? = HomeToWorkClient.user

        if (user != null) {
            callback.onValidSession()
            return
        }

        val token = prefs.getString(PREFS_TOKEN, "")

        if (token.isEmpty()) {
            callback.onInvalidSession(Const.CODE_NO_ACCESS_TOKEN, null)
            return
        }

        HomeToWorkClient.login(token, object : AuthCallback {
            override fun onSuccess() {

                // Crashlytics log
                Answers.getInstance().logLogin(LoginEvent()
                        .putMethod("Token")
                        .putSuccess(true))

                Crashlytics.setUserIdentifier(HomeToWorkClient.user?.id.toString())
                Crashlytics.setUserEmail(HomeToWorkClient.user?.email)
                Crashlytics.setUserName(HomeToWorkClient.user.toString())

                callback.onValidSession()

            }

            override fun onInvalidCredential() {
                clearSession(ctx)
                callback.onInvalidSession(Const.CODE_INVALID_CREDENTIALS, null)
            }

            override fun onError(throwable: Throwable?) {

                // Crashlytics log
                Answers.getInstance().logLogin(LoginEvent()
                        .putMethod("Token")
                        .putSuccess(false))

                if (throwable is UnknownHostException) {
                    callback.onInvalidSession(Const.CODE_SERVER_ERROR, throwable)
                } else {
                    callback.onInvalidSession(Const.CODE_LOGIN_ERROR, null)
                }

            }
        })
    }

    fun clearSession(ctx: Context) {
        ctx.stopService(Intent(ctx, LocationService::class.java))
        ctx.stopService(Intent(ctx, SyncService::class.java))

        val prefs = ctx.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE)
        val editor = prefs.edit()
        editor.clear()
        editor.apply()
    }

    interface SessionCallback {
        fun onValidSession()
        fun onInvalidSession(code: Int, throwable: Throwable?)
    }

}
